src/sorted_container_iteration.jl

```diff
@@ -0,0 +1,1167 @@
+{C, KV, T, D} = SizeUnknown()
+
+Base.length(ito::IterableObject{C, EntireContainer, KV, T, D}) where
+{C, KV, T, D} = length(ito.m)
+
+
+"""
+    Base.in(x, iter::SortedContainerIterable)
+
+Returns true if `x` is in `iter`, where `iter` refers to any of the
+iterable objects described under [`Base.iterate(iter::SortedContainerIterable)`](@ref),
+and `x` is of the appropriate type. For all of the iterables
+except the five listed below, the algorithm used is a linear-time
+search. For example, the call:
+
+    (k=>v) in exclusive(sd, st1, st2)
+
+where `sd` is a SortedDict, `st1` and `st2` are semitokens, `k` is a
+key, and `v` is a value, will loop over all entries in the
+dictionary between the two tokens and a compare for equality using
+`isequal` between the indexed item and `k=>v`.
+
+The five exceptions are:
+
+```julia
+(k=>v) in sd
+(k=>v) in smd
+k in ss
+k in keys(sd)
+k in keys(smd)
+```
+
+Here, `sd` is a SortedDict, `smd` is a SortedMultiDict, and `ss` is
+a SortedSet.
+
+These five invocations of `in` use the index structure of the sorted
+container and test equality based on the order object of the keys
+rather than `isequal`. Therefore, these five are all faster than
+linear-time looping. To force the use of `isequal` test on
+the keys rather than the order object (thus slowing the execution
+from logarithmic to linear time), replace the above five constructs
+with these:
+
+```julia
+(k=>v) in collect(sd)
+(k=>v) in collect(smd)
+k in collect(ss)
+k in collect(keys(sd))
+k in collect(keys(smd))
+```
+"""
+Base.in(x, m::SortedContainerIterable) =
+    invoke(in, Tuple{Any,Any}, x, m)
+
+Base.in(k, ito::IterableObject{C, EntireContainer, KeysIter, NoTokens, D}) where
+{C, D} = haskey(ito.m, k)
+
+
+"""
+    haskey(sc::SortedContainer, k)
+
+Return `true` iff key `k` is present in `sc`.  Equivalent
+to 
+`in(k,sc)` for a SortedSet, or to `in(k,keys(sc))` for
+a SortedDict or SortedMultiDict.  Time: O(*c* log *n*)
+"""
+@inline function Base.haskey(m::SortedContainer, k_) 
+    i, exactfound = findkey(m.bt, convert(keytype(m),k_))
+    return exactfound
+end
+
+
+Base.eachindex(sd::SortedDict) = keys(sd)
+Base.eachindex(sdm::SortedMultiDict) = onlysemitokens(sdm)
+Base.eachindex(ss::SortedSet) = onlysemitokens(ss)
+
+
+Base.eachindex(ito::IterableObject{C, R, KeysValsIter, NoTokens, D}) where
+{C <: SortedDict, R, KV, D} = keys(ito)
+
+Base.eachindex(ito::IterableObject{C, R, KV, NoTokens, D}) where
+{C <: SortedMultiDict, R, KV, T, D} = onlysemitokens(ito)
+
+Base.eachindex(ito::IterableObject{C, R, KV, NoTokens, D}) where
+{C <: SortedSet, R, KV, T, D} = onlysemitokens(ito)
+
+
+"""
+   empty!(m)
+
+Empty a sorted container
+"""
+
+Base.empty!(m::SortedContainer) = (empty!(m.bt); m)
+Base.length(m::SortedContainer) = length(m.bt.data) - length(m.bt.freedatainds) - 2
+Base.isempty(m::SortedContainer) = length(m) == 0
+
+
+(:)(t1::Token, t2::Token) = _colon(t1,t2)
+
+function _colon(t1::Token, t2::Token)
```

Comment:
why not define this as `:` ?

Reply:
The definition `function Base.(:)(a,b)` gives a syntax error (noted above).  However, `function (:)(a,b)` works, so I have fixed this (but I still have the issue with `import` noted above).

Reply:
OK, I figured this out.